### DISCUSSION TITLE
`array_api` usage

### DISCUSSION THREAD

**Discussion by ilan-gold** [zarr-developers/zarr-python#2197] [Q&A]
Hello all,

I have been looking into the core code here (teehee) and I noticed the recent #1967.  It made me wonder why `array_api` + dlpack is not used (in general, and also specifically in that PR).  I am guessing "no one has done it yet" is the answer, which is fair considering we all have other things going on, but then I am curious what #1967 gives that `array_api` + dlpack wouldn't.

I think `cupy`, `jax` (mentioned in the PR), and `numpy` all support `array_api` if memory serves correctly.

Sorry for the short question, with probably a long answer, but don't want to presume too much.  I also did a quick search of the issues and found nothing relevant to the array api, but might have missed something: https://data-apis.org/array-api/latest/index.html

**Comment by d-v-b** [comment]
We discussed the array API when we were thinking about the design for the Zarr array class. See https://github.com/zarr-developers/zarr-python/discussions/1614. My thoughts today are pretty much the same as they were back then -- the array api describes a lot of transformations of in-memory arrays, but zarr arrays are not generally in-memory, which puts a lot of the array API off limits for us without basically re-inventing dask. 

That doesn't mean we should deliberately deviate from the array api of course. But I don't think we are the target for that standard.

**Reply by dcherian** [reply]
I think the key part of the question is `dlpack` rather than array-api. But I *think* the answer there is that `dlpack` allows converting after the data is read, and the `buffer` argument controls the "in-memory" format of the bytes being read from storage.

**Reply by ilan-gold** [reply]
@d-v-b I meant more for the internal `Buffer` class - apologies.  So using it internally as a check for in-memory representation choice.  It seemed like #1967 was a lot of code, but a lot of the internal `Buffer` stuff on both `gpu` and `cpu` seems array-api compatible?   For example I see `cp.asarray` (in `array_api`), `cp.array` (which I think is (a) deriveable from some reference, which would be needed for `array_api` and (b) replaceable with something `array_api` compatible in most spots), `cp.frombuffer` (very similar to `dlpack`, I think), `cp.concatenate`(in `array_api`), `cp.empty` (in `array_api`)....and so forth.  We can then special case from there if necessary.  This would allow `jax` as well.  Maybe I'm misreading your original post though.

**Reply by d-v-b** [reply]
this clarification is helpful, I think I'm the one who misunderstood things :)  maybe @akshaysubr and @madsbk can shed more light on the design of the `Buffer` class in the context of your question?

**Reply by ilan-gold** [reply]
Thanks!  And just to clarify with @dcherian's point - the `array_api` _requires_ `dlpack`.  We could get it for free basically, see e.g., https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html#dlpack

**Reply by ilan-gold** [reply]
I think I will convert this into an issue since I am now a bit more convinced of its utility.

**Reply by ilan-gold** [reply]
Also @d-v-b I am 100% in agreement.  It does not make sense for the `Array` class.

**Reply by akshaysubr** [reply]
Thanks for bringing this up. This is certainly worth some discussion and in fact, DLPack was in consideration but just not prioritized for the initial implementation. I'll comment a bit more on the issue that's created to avoid fragmentation.
